sdk/nodejs/docker.ts

```diff
@@ -186,32 +186,28 @@ async function buildAndPushImageWorkerAsync(
 
     const { imageName, tag } = utils.getImageNameAndTag(baseImageName);
 
-    let loggedIn: Promise<void> | undefined;
-
-    // If no `connectToRegistry` function was passed in we simply assume docker is already logged-in to the correct registry (or uses auto-login via credential helpers)
-    // hence we resolve loggedIn immediately.
-    if (!connectToRegistry) {
-        loggedIn = Promise.resolve();
-    }
-    const login = () => {
-        if (!loggedIn) {
+    const pullFromCache = typeof pathOrBuild !== "string" && pathOrBuild && pathOrBuild.cacheFrom && !!repositoryUrl;
+    if (!pulumi.runtime.isDryRun() || pullFromCache) {
```

Comment:
we only had two paths that could lead to logging in.  one was once we built the image and wanted to push (which only happens in *update* and not *preview*.  The other is if we needed to pull cache layers.  So we jsut check if either of those will be happening, and we login up front once.  

This also made things much simpler as it meant not having these contortions to log in when necessayr, but only login once.

Reply:
Worth putting a short comment on `pullFromCache` to make clear the cases it is trying to capture?

Reply:
sure :)